odoo/addons/base/models/res_lang.py

```diff
@@ -153,22 +178,238 @@ def fix_datetime_format(format):
+        if not module:
+            return None
+
+        attachment_name = "{}/i18n/{}.po".format(addons, lang)
+        po_file = self.env['ir.attachment'].search([
+            ('res_model', '=', 'ir.module.module'),
+            ('res_id', '=', module.id),
+            ('name', '=', attachment_name)
+        ], limit=1)
+
+        if po_file:
+            yield StringIO(po_file.datas)
+
+    @api.model
+    def _extract_po_to_i18n_folder(self, addons, tmp, extracted_file):
+        """ Extraction method if the po are stored on filesystem, inside /i18n/ folders """
+        path = get_resource_path(addons)
+        if not os.access(path, os.W_OK):
+            _logger.warning("Path %s is not writeable", path)
+            return False
+
+        src_file = os.path.join(tmp, extracted_file)
+        dst_file = os.path.join(path, extracted_file)
+        shutil.move(src_file, dst_file)
+
+    @api.model
+    def _extract_po_to_attachment(self, addons, tmp, extracted_file):
+        """ Extraction method if the po are stored in db, inside ir.attachments """
+        module = self.env['ir.module.module'].search([('name', '=', addons)], limit=1)
+        if not module:
+            _logger.info("Skip translations extraction for unexpected module %s", addons)
+            return
+
+        src_file = os.path.join(tmp, extracted_file)
+        with open(src_file, 'r') as po_file:
+            po_content = base64.b64encode(po_file.read().encode()).decode()
+
+        # sudo to be able to search and write, even if no rights on ir.module.module
+        Attachment = self.env['ir.attachment'].sudo()
+
+        attached_po = Attachment.search([
+            ('res_model', '=', 'ir.module.module'),
+            ('res_id', '=', module.id)
+            ('name', '=', extracted_file)
+        ], limit=1)
+        if attached_po:
+            attached_po.write({'datas': po_content})
+        else:
+            attached_po = Attachment.create({
+                'datas': po_content,
+                'res_model': 'ir.module.module',
+                'res_id': module.id,
+                'name': extracted_file,
+            })
+        return attached_po
+
+    @api.model
+    def _extract_i18n_file_content(self, fileobj, lang, module_list, extraction_method):
+        """ Extract the translations from the given archive
+
+        The expected archive is a .tar.xz file using the structure:
+            <module>/i18n/<lang>.po
+
+        Regional variants are accepted (e.g. fr.tar.xz can contain both fr and
+        fr_BE files)
+
+        :param fileobj: a file object containing the compressed translation files
+        :param lang: the simplified language code to load (e.g. 'fr')
+        :param module_list: the list of modules to update from this archive
+        :param extraction_method: the method to call with the extracted .po file
+        """
+        with tarfile.open(mode='r:xz', fileobj=fileobj) as tar_content:
```

Comment:
Maybe we could avoid being super prescriptive, use `r`/`r:*` here and not put an extension in the lookup URL e.g. just fetch `/i18n/12.0/en_US` under the assumption that it's a tarfile and nothing more? (technically people could use a plain tar with content-encoding: gzip served by an nginx and that'd work fine for us).